File: src/model/Parser.cpp
//
// Created by vkolobara on 12/24/17.
//

#include "Parser.h"

using namespace std;

void Parser::parse(string filePath) {
    ifstream file(filePath);

    string line;

    getline(file, line);

    stringstream iss(line);

    double maxWalk;
    unsigned int capacity;
    unsigned int nStops;
    unsigned int nStudents;
    string garbage;

    // Read numbers
    iss >> nStops >> garbage >> nStudents >> garbage >> maxWalk >> garbage >> garbage >> capacity >> garbage;
    this->maxWalk = maxWalk;
    busCapacity = capacity;

    const unsigned int rows = nStudents;
    const unsigned int cols = nStops;

    // Skip empty line
    getline(file, line);

    // Read school
    double x,y;
    int id;

    getline(file, line);
    iss.clear();
    iss.str(line);

    iss >> id >> x >> y;
    school = make_shared<Coordinate>(Coordinate(x,y));

    nStops--;
    while(nStops--) {
        getline(file, line);
        stringstream ss(line);

        ss >> id >> x >> y;
        Coordinate coord(x,y);
        busStops.emplace_back(coord);
        stopToSchoolDistance.emplace_back(Coordinate::euclideanDistance(*school.get(), coord));
    }

    getline(file, line);
    getline(file, line);

    while(nStudents--) {
        getline(file, line);
        stringstream ss(line);
        ss >> id >> x >> y;
        Coordinate coord(x,y);
        students.emplace_back(coord);

        vector<double> distances;

        for (int i=0; i<busStops.size(); i++) {
            double dst = Coordinate::euclideanDistance(coord, busStops[i]);
            distances.push_back(dst);
        }

        studentToStopDistance.push_back(distances);

    }
    file.close();

}

Parser::Parser() {

}

unsigned int Parser::getBusCapacity() const {
    return busCapacity;
}

double Parser::getMaxWalk() const {
    return maxWalk;
}

const shared_ptr<Coordinate> &Parser::getSchool() const {
    return school;
}

const vector<Coordinate> &Parser::getBusStops() const {
    return busStops;
}

const vector<Coordinate> &Parser::getStudents() const {
    return students;
}

Parser::~Parser() {
}

const vector<vector<double>> &Parser::getStudentToStopDistance() const {
    return studentToStopDistance;
}

const vector<double> &Parser::getStopToSchoolDistance() const {
    return stopToSchoolDistance;
}

File: src/hr/vinko/hmo/parser/Parser.java
package hr.vinko.hmo.parser;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class Parser {
	private Coordinate school;
	private List<Coordinate> busStops;
	private List<Coordinate> students;

	private int busCapacity;
	private double maxWalk;

	private double[][] studentToStopDistance;
	private double[] stopToSchoolDistance;
	
	public Parser(int nStops, int nStudents) {
		busStops = new ArrayList<>(nStops);
		students = new ArrayList<>(nStudents);
		studentToStopDistance = new double[nStudents][nStops];
		stopToSchoolDistance = new double[nStops-1];
	}

	public static Parser parse(String filePath) throws IOException {
		try (Scanner scanner = new Scanner(new File(filePath))) {
			String[] line = scanner.nextLine().split("\\s+");
			int nStops = Integer.parseInt(line[0]);
			int nStudents = Integer.parseInt(line[2]);
			
			Parser parser = new Parser(nStops, nStudents);
			parser.maxWalk = Double.parseDouble(line[4]);
			parser.busCapacity = Integer.parseInt(line[7]);
			
			scanner.nextLine();
			
			line = scanner.nextLine().split("\\s+");
			parser.school = new Coordinate(Double.parseDouble(line[1]), Double.parseDouble(line[2]));
			
			
			for (int i=0; i<nStops-1; i++) {
				line = scanner.nextLine().split("\\s+");
				Coordinate busStop = new Coordinate(Double.parseDouble(line[1]), Double.parseDouble(line[2]));
				parser.busStops.add(busStop);
				parser.stopToSchoolDistance[i] = Coordinate.euclideanDistance(busStop, parser.school);
			}
			
			scanner.nextLine();
			scanner.nextLine();
			
	
			
			for (int i=0; i<nStudents; i++) {
				line = scanner.nextLine().split("\\s+");
				Coordinate student = new Coordinate(Double.parseDouble(line[1]), Double.parseDouble(line[2]));
				parser.students.add(student);
				
				for (int j=0; j<nStops-1; j++) {
					parser.studentToStopDistance[i][j] = Coordinate.euclideanDistance(student, parser.busStops.get(j));
				}
			}
			
			return parser;

		} catch (IOException e) {
			e.printStackTrace();
		}
		
		return null;
	}

	public Coordinate getSchool() {
		return school;
	}

	public List<Coordinate> getBusStops() {
		return busStops;
	}

	public List<Coordinate> getStudents() {
		return students;
	}

	public int getBusCapacity() {
		return busCapacity;
	}

	public double getMaxWalk() {
		return maxWalk;
	}

	public double[][] getStudentToStopDistance() {
		return studentToStopDistance;
	}

	public double[] getStopToSchoolDistance() {
		return stopToSchoolDistance;
	}

	

}
